api/admin/controller/patron_auth_services.py

```diff
@@ -43,6 +49,30 @@ def __init__(self, manager):
                                ]
         self.type = _("patron authentication service")
 
+    def _get_validator(self, protocol, integration):
+        """
+        Returns a configured validator for a specific protocol
+
+        :param protocol: Patron authentication protocol
+        :type protocol: string
+
+        :param integration: Integration
+        :type integration: ExternalIntegration
+
+        :return: Validator object (if there is any for the specific protocol)
+        :rtype: Optional[Validator]
+        """
+
+        if protocol == 'api.saml.provider':
+            saml_settings_validator = SAMLSettingsValidator(
+                SAMLMetadataParser(),
+                SAMLMetadataSerializer(integration)
+            )
+
+            return saml_settings_validator
+
+        return None
+
```

Comment:
I was thinking about registering validators inside `LibraryAuthenticator.register_provider`, storing registered inside `LibraryAuthenticator._validators_by_protocol` dictionary and accessing it via `Authenticator.validator_lookup`, `LibraryAuthenticator.validator_lookup`.
For example, `Authenticator.validator_lookup` was supposed to look like:
```python
    def validator_lookup(self, *args, **kwargs):
        return self.invoke_authenticator_method(
            "validator_lookup", *args, **kwargs
        )
```

But it turned out that I can't use it that way because there is no library associated with the current request during settings validation in `PatronAuthServicesController`.

@leonardr, is it the right way to implement it?

Reply:
`LibraryAuthenticator` is designed for situations where the request is scoped to a specific library, so that won't work. And `Authenticator` is a collection of `LibraryAuthenticator` objects.

What if you defined a function called `validator_factory` in api.saml.provider which returned a `SAMLSettingsValidator` method when called with an ExternalIntegration? This method could then look for a `validator_factory` in the appropriate module and call it if it finds one.

Reply:
I created `ValidatorFactory` class inside `api.admin.valiator` and used it in `PatronAuthServicesController`